providers/providers.go

```diff
@@ -191,7 +195,22 @@ func writeProviderEntry(dstore ds.Datastore, k cid.Cid, p peer.ID, t time.Time)
 	return dstore.Put(ds.NewKey(dsk), buf[:n])
 }
 
-func (pm *ProviderManager) gc() {
+func (pm *ProviderManager) gc(proc goprocess.Process) {
+	// Don't use the batching dstore, it's not thread-safe and we call this
+	// in a separate thread.
+
+	// Note 1: we'll ignore some puts/deletes buffered in the batching
+	// datastore. Basically, this means we may not GC something we should be
+	// GCing but, meh.
+	//
+	// Note 2: there's a tiny race where we may remove a non-expired provider record:
+	//
+	// 1. Record expires.
+	// 2. We start a query for the record.
+	// 3. New put comes through for the record.
+	//
+	// While fixable, I'm not sure it's worth it as the DHT isn't supposed
+	// to have reliable storage anyways.
```

Comment:
It's probably worth fixing, as a provide won't be republished for hours potentially.

Reply:
This only happens if someone republishes immediately after the provider has expired. The alternative is to keep track of all puts while GCing. I'll try but it's going to suck for minimal benefit.

Reply:
If it turns out it's too ugly, we can drop it.

Reply:
I'm going to try.